src/commands/run.rs

```diff
@@ -370,12 +396,49 @@ impl RunCommand {
         result
     }
 
-    fn load_main_module(
-        &self,
-        store: &mut Store<Host>,
-        linker: &mut Linker<Host>,
-        module: Module,
-    ) -> Result<()> {
+    fn setup_epoch_handler(&self, store: &mut Store<Host>, module_name: &str) -> Box<dyn FnOnce()> {
+        if let Some(path) = &self.profile_guest {
+            let interval = self.profile_guest_interval;
+            let profiler = Arc::new(Mutex::new(GuestProfiler::new(module_name, interval)));
```

Comment:
API question: instead of requiring an `Arc` and `Mutex` here, which technically aren't necessary since there aren't any threads, could this instead be stored in the `T` of `Store<T>`? As-is the `GuestProfiler::sample` method would not work since it would be stored in `T` but additionally takes a store as a parameter, but I think that could be fixed by taking a`&WasmBacktrace` as an argument since that's all the store is used for internally.

Reply:
"Wait, I could just keep it in the Store," I thought, as I fell asleep last night; then wake up to find this comment. :laughing: 

I don't want to make `WasmBacktrace` part of the public API here because I want to re-implement this in terms of the lower-level `wasmtime-runtime` stack-walking API. That doesn't directly need access to the store at all, but I do need some information from the store at some point to figure out how native PCs map to individual modules and to functions within those modules.

It might work though to pass in the store when constructing a `GuestProfiler`, collect just the necessary information about all the modules and symbols from it, and then not need it in `sample`.

Otherwise, the only alternative I've thought of is to define `fn sample(store: impl AsContextMut, get_self: impl FnOnce(&mut Store) -> &mut Self)`. Not the nicest interface but I think it should work, right?

Reply:
How about:

```rust
impl<T> Store<T> {
    pub fn sample(&mut self, get_profiler: impl FnOnce(&mut T) -> &mut GuestProfiler) { ... }
}
```

with similar impls for `StoreContextMut`?

which is more-or-less the same thing, but just as a method instead of a function on `GuestProfiler` now that I reread your comment